pkg/flist/cleanup.go

```diff
@@ -0,0 +1,140 @@
+package flist
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+// Cleaner interface, implementer of this interface
+// can start a cleaner job
+type Cleaner interface {
+	// Cleaner runs the clean process, Cleaner should be
+	// blocking. Caller then can do `go Cleaner()` to run it in the background
+	Cleaner(ctx context.Context, every time.Duration)
+}
+
+var _ Cleaner = (*flistModule)(nil)
+
+func (f *flistModule) listMounts() (map[string]int64, error) {
+	infos, err := ioutil.ReadDir(f.run)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to list g8ufs process pids: %s", f.pid)
+	}
+
+	data := make(map[string]int64)
+	for _, info := range infos {
+		if info.IsDir() {
+			// what is that doing here?
+			continue
+		}
+
+		path := filepath.Join(f.run, info.Name())
+		pid, err := f.getPid(path)
+		if os.IsNotExist(err) {
+			pid = -1
+		} else if err != nil {
+			log.Error().Err(err).Str("pid-file", path).Msg("invalid pid file")
+		}
+
+		data[info.Name()] = pid
+	}
+
+	return data, nil
+}
+
+// cleanupMount forces clean up of a mount point
+func (f *flistModule) cleanupMount(name string) error {
+	path, err := f.mountpath(name)
+	if err != nil {
+		return err
+	}
+
+	// always clean up files
+	defer func() {
+		for _, file := range []string{
+			filepath.Join(f.pid, fmt.Sprintf("%s.pid", name)),
+			filepath.Join(f.run, name),
+			filepath.Join(f.log, fmt.Sprintf("%s.log", name)),
+		} {
+			if err := os.Remove(file); err != nil {
+				log.Warn().Err(err).Str("file", file).Msg("failed to delete pid file")
+			}
+		}
+	}()
+	// the following procedure will ignore errors
+	// because the only point of this is to force
+	// clean the entire mount.
+	// so no checks or error checking is done
+
+	if err := syscall.Unmount(path, syscall.MNT_DETACH); err != nil {
+		log.Warn().Err(err).Str("path", path).Msg("fail to unmount flist")
+	}
+
+	volume, err := f.storage.Path(name)
+	if err != nil {
+		log.Warn().Err(err).Str("subvolume", name).Msg("subvolume does not exist")
+		return nil
+	}
+
+	ro := filepath.Join(volume, "ro")
```

Comment:
we only need to umount the ro layer ? what about the others? Or do we assume overlayfs always clean up properly ?

Reply:
There are no others, there is only the "mountpoint" which is fuse, which is unmounted on line 77. then only the fuse is mounted under the "ro". the other directories "wd", and "rw" are just folders with data in them

Reply:
I don't think that's true. overlayfs is also mounted by 0-fs. So there are 2 mounted filesystem:
- 0-fs which provide the ro layer
- overlays which user the ro and mount on `mountpoint`

What about overlayfs ? we don't deal with it ?

Reply:
Overlayfs is only mounted if you mount the fs in read-write mode. because that's it's use case, if you start the fs in read-only mode it only mount the fuse directly under your target mountpoint.

Reply:
I think I get your question now, the `ro` directory is the only mounted path under the backend, the overlay fs is mounted under the mount target that has been unmounted already. 
In some cases (ro mount of 0fs) the fuse is mounted directly under the target, hence nothing to do under the volume (it will not exist already)